src/ripple/nodestore/impl/DatabaseShardImp.cpp

```diff
@@ -392,15 +414,28 @@ DatabaseShardImp::importShard(
             return false;
         }
 
-        dstDir = dir_ / std::to_string(shardIndex);
+        auto const expectResult =
+            expectNewShard(shardIndex, countHistoricalShards(lock), lock);
+
+        if (!expectResult)
+        {
+            JLOG(j_.error()) << "shard " << shardIndex << " failed to import";
+            return false;
+        }
+
+        auto const needsHistoricalPath = *expectResult;
+        dstDir = needsHistoricalPath ? chooseHistoricalPath(lock) : dir_;
     }
 
     // Rename source directory to the shard database directory
-    if (!renameDir(srcDir, dstDir))
+    if (!renameDir(srcDir, dstDir / std::to_string(shardIndex)))
         return false;
 
     // Create the new shard
-    auto shard{std::make_unique<Shard>(app_, *this, shardIndex, j_)};
+    auto shard{std::make_unique<Shard>(app_, *this, shardIndex, j_, dstDir)};
+
+    dstDir /= std::to_string(shardIndex);
```

Comment:
We convert `shardIndex` to a string in two places. We may want to create a local var for this (we do it here and line 431).

Reply:
Done